### PR TITLE
feat(types): Introduce SafePromiseLike

### PR DESCRIPTION
**Pull request by rozzilla**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Address https://github.com/fastify/fastify/issues/5498

### CODE REVIEW

File: test/types/fastify.test-d.ts [hunk 1]
```diff
@@ -11,30 +11,32 @@ import fastify, {
...
-  FastifyError
+  FastifyError,
+  SafePromiseLike
 } from '../../fastify'
 import { ErrorObject as AjvErrorObject } from 'ajv'
 import * as http from 'http'
 import * as https from 'https'
 import * as http2 from 'http2'
 import { expectType, expectError, expectAssignable, expectNotAssignable } from 'tsd'
-import { FastifyLoggerInstance } from '../../types/logger'
```
Comment (rozzilla) [inline test/types/fastify.test-d.ts:21]: Cleanup, since it's no longer used 🧹

File: docs/Reference/TypeScript.md [hunk 2]
```diff
@@ -661,6 +661,21 @@ However, there are a couple of suggestions to help improve this experience:
 - Make sure the `no-unused-vars` rule is enabled in
   [ESLint](https://eslint.org/docs/rules/no-unused-vars) and any imported plugin
   are actually being loaded.
+- In case you've the `@typescript-eslint/no-floating-promises` enabled,
+please double-check that your `.eslintrc` includes `allowForKnownSafePromises`
+property as described [here](https://github.com/fastify/fastify/issues/5498):
```
Comment (mcollina) [inline docs/Reference/TypeScript.md:666]: Can we point to the official docs instead?

File: docs/Reference/TypeScript.md [hunk 3]
```diff
@@ -661,6 +661,21 @@ However, there are a couple of suggestions to help improve this experience:
...
+```
+{
+  "rules": {
+    "@typescript-eslint/no-floating-promises": ["error", {
+      "allowForKnownSafePromises": [
+        // { "from": "package", "name": "FastifySafePromiseOrSomeSuch", "package": "@fastify/etc" },
+        { "from": "file", "name": "SafePromise"}
```
Comment (mcollina) [inline docs/Reference/TypeScript.md:673]: Does this capture all of Fastify?

File: docs/Reference/TypeScript.md [hunk 4]
```diff
@@ -661,6 +661,21 @@ However, there are a couple of suggestions to help improve this experience:
...
   are actually being loaded.
+- In case you've the `@typescript-eslint/no-floating-promises` enabled,
+please double-check that your `.eslintrc` includes `allowForKnownSafePromises`
+property as described [here](https://github.com/fastify/fastify/issues/5498):
+```
+{
+  "rules": {
+    "@typescript-eslint/no-floating-promises": ["error", {
+      "allowForKnownSafePromises": [
+        // { "from": "package", "name": "FastifySafePromiseOrSomeSuch", "package": "@fastify/etc" },
```
Comment (mcollina) [inline docs/Reference/TypeScript.md:672]: Can this be removed? Alternatively, can you specify an actual plugin/real case?

**Review by mcollina — COMMENTED** [review]
@JoshuaKGoldberg can you take a look?

Our end goal would be to have users have to specify this in the first place and get this config from us.

File: types/type-provider.d.ts [hunk 5]
```diff
@@ -93,3 +93,11 @@ RouteGeneric
...
 // (Promise<Return> | Return | Promise<void> | void)
   : unknown
+
+/**
+ * This branded type is needed to please `typescript-eslint`.
+ *
+ * Please refer to the following Github issue for more info:
+ * https://github.com/fastify/fastify/issues/5498
+ */
+export type SafePromiseLike<T> = PromiseLike<T> & { __brand: 'SafePromiseLike' }
```
Comment (climba03003) [inline types/type-provider.d.ts:103]: I want to ensure we didn't use the wrong property.
It is a bit confuse to me `__brand` is not showing in the document, and `redux-tool` is using `__linterBrands`

https://github.com/fastify/fastify/issues/5498#issuecomment-2154566979

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: docs/Reference/TypeScript.md (hunk 2 above)
Comment (rozzilla) [reply docs/Reference/TypeScript.md]: Done on [52c9dab](https://github.com/fastify/fastify/pull/5506/commits/52c9dab1eb4537752aa882acebbbb8187c2c6c9b)

File: docs/Reference/TypeScript.md (hunk 3 above)
Comment (rozzilla) [reply docs/Reference/TypeScript.md]: Updated on [52c9dab](https://github.com/fastify/fastify/pull/5506/commits/52c9dab1eb4537752aa882acebbbb8187c2c6c9b)

File: docs/Reference/TypeScript.md (hunk 4 above)
Comment (rozzilla) [reply docs/Reference/TypeScript.md]: Removed on [52c9dab](https://github.com/fastify/fastify/pull/5506/commits/52c9dab1eb4537752aa882acebbbb8187c2c6c9b)

File: types/type-provider.d.ts (hunk 5 above)
Comment (rozzilla) [reply types/type-provider.d.ts]: Good point @climba03003. On the [eslint](https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafepromises) doc they recommend using `__linterBrands`:
image

Updated on [cb38496](https://github.com/fastify/fastify/pull/5506/commits/cb3849612b128fcd2b90da8578e2055f85366d46) ✅

File: types/type-provider.d.ts [hunk 6]
```diff
@@ -93,3 +93,11 @@ RouteGeneric
 // review: support both async and sync return types
 // (Promise<Return> | Return | Promise<void> | void)
   : unknown
+
+/**
+ * This branded type is needed to please `typescript-eslint`.
```
Comment (JoshuaKGoldberg) [inline types/type-provider.d.ts:98]: I protest this phrasing 😄. typescript-eslint is just one tool for typed linting that happens to be the only popular one right now. Soon, others such as Biome, deno lint, and oxc will have equivalent `no-floating-promises` rules. Those rules will likely by default also report on the same situations that [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises) reports.

In other words, this `SafePromiseLike` is needed _for the code to explain itself_, not because any one particular external tool has its own opinions.

```suggestion
 * This branded type is needed to indicate APIs that return Promise-likes which can
 * safely "float" (not have rejections handled by calling code).
```

**Review by climba03003 — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
LGTM

File: types/type-provider.d.ts (hunk 6 above)
Comment (rozzilla) [reply types/type-provider.d.ts]: Updated on [0b9de79](https://github.com/fastify/fastify/pull/5506/commits/0b9de798b62f53a6cebe09e4272ca3906efcf905) ✅

File: docs/Reference/TypeScript.md [hunk 7]
```diff
@@ -661,6 +661,20 @@ However, there are a couple of suggestions to help improve this experience:
...
   are actually being loaded.
+- In case you've the `@typescript-eslint/no-floating-promises` enabled,
+please double-check that your `.eslintrc` includes `allowForKnownSafePromises`
+property as described [here](https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafepromises):
+```
+{
+  "rules": {
+    "@typescript-eslint/no-floating-promises": ["error", {
+      "allowForKnownSafePromises": [
+        { "from": "package", "name": "SafePromiseLike", "package": "fastify" }
```
Comment (JoshuaKGoldberg) [inline docs/Reference/TypeScript.md:672]: I set up a reproduction here: https://github.com/JoshuaKGoldberg/repros/tree/fastify-no-floating-promises-repros

...but I couldn't find any cases where `PromiseLike` types are exposed to users. Out of the two issues with repro cases I could find from [`is:issue no-floating-promises`](https://github.com/fastify/fastify/issues?q=is%3Aissue+no-floating-promises):
* #4246: triggers on `reply.header()`, which returns `FastifyReply`
* #4399: triggers on `reply.send()`, which returns `FastifyReply`

In fact, there's no reference to `PromiseLike` in `fastify/types/reply.d.ts` at all. Most methods return `FastifyReply`.

So, can we get away with just recommending using `allowForKnownSafePromises` on `FastifyReply`, and not making any `.d.ts` changes?

```suggestion
        { "from": "package", "name": "FastifyReply", "package": "fastify" }
```

(or: are there cases where `PromiseLike` is indeed exposed to users?)

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Thanks for working on this! I can't find a way to (`Safe`)`PromiseLike` show up for end-user code, so - maybe this is "just" a docs change?

File: docs/Reference/TypeScript.md (hunk 7 above)
Comment (mcollina) [reply docs/Reference/TypeScript.md:672]: `PromiseLike` is exposed via https://github.com/fastify/fastify/blob/c33cc90802b46570139430999cf290b85feb1856/types/instance.d.ts#L173. We use that technique instead of declaring a `.then()` method in `FastifyInstance` because typescript-eslint would then require the users to add `await` everywhere.

Something that this is not addressing is:
 https://github.com/fastify/fastify/blob/c33cc90802b46570139430999cf290b85feb1856/types/reply.d.ts#L77 and

Comment (JoshuaKGoldberg) [reply docs/Reference/TypeScript.md:673]: In the meantime:

```suggestion
        { "from": "package", "name": "FastifyReply", "package": "fastify" },
        { "from": "package", "name": "SafePromiseLike", "package": "fastify" },
```

Comment (rozzilla) [reply docs/Reference/TypeScript.md]: @JoshuaKGoldberg  should we also add `FastifyInstance` to the list?

Comment (JoshuaKGoldberg) [reply docs/Reference/TypeScript.md]: I don't see a `.then()` for `FastifyInstance`, do you have a repro case showing it triggering `no-floating-promises`?

Comment (rozzilla) [reply docs/Reference/TypeScript.md]: Yes, in this case the behavior it's extremely weird.
```
fastify(); // cause the error no-floating-promises
const something = fastify(); // no error
```
So assigning `fastify` to a variable, avoid the `eslint` error 🤔

Comment (JoshuaKGoldberg) [reply docs/Reference/TypeScript.md]: Now I'm curious about the API design: why would one not assign `fastify()` / a `FastifyInstance` somewhere? And/or, why does it have a `.then()`?

Comment (rozzilla) [reply docs/Reference/TypeScript.md]: @mcollina know for sure :)

Comment (mcollina) [reply docs/Reference/TypeScript.md]: > And/or, why does it have a .then()?

It does have a `then()` so that it start loading all plugins. We are not typing it because otherwise _every single method_ in Fastify would need to be awaited according to typescript-eslint and users have been complaining about that. In other terms, a user can do `await app.register(myPlugin)` instead of `app.register(myPlugin); await app.ready()`. It's syntactic sugar that improves the developer experience.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
@JoshuaKGoldberg can you do a final pass on this? Or are we waiting for a fix on typescript-eslint side?

File: docs/Reference/TypeScript.md (hunk 7 above)
Comment (JoshuaKGoldberg) [reply docs/Reference/TypeScript.md]: > In other terms, a user can do `await app.register(myPlugin)` instead of `app.register(myPlugin); await app.ready()`.

Right, thanks, this is helpful - I hadn't seen folks doing this (in part because I don't often work in the areas of code Fastify is made for). I'll mention in https://github.com/typescript-eslint/typescript-eslint/issues/8433 too.

File: docs/Reference/TypeScript.md [hunk 8]
```diff
@@ -661,6 +661,22 @@ However, there are a couple of suggestions to help improve this experience:
...
+- In case you've the `@typescript-eslint/no-floating-promises` enabled,
+please double-check that your `.eslintrc` includes `allowForKnownSafePromises`
+property as described on the [`typescript-eslint no-floating-promises allowForKnownSafePromises
+documentation`](https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafepromises):
+```
+{
+  "rules": {
+    "@typescript-eslint/no-floating-promises": ["error", {
+      "allowForKnownSafePromises": [
+        { "from": "package", "name": "FastifyReply", "package": "fastify" },
```
Comment (JoshuaKGoldberg) [inline docs/Reference/TypeScript.md:673]: Per https://github.com/fastify/fastify/pull/5506#discussion_r1631165436:

```suggestion
        { "from": "package", "name": "FastifyInstance", "package": "fastify" },
        { "from": "package", "name": "FastifyReply", "package": "fastify" },
```

File: docs/Reference/TypeScript.md [hunk 9]
```diff
@@ -661,6 +661,22 @@ However, there are a couple of suggestions to help improve this experience:
 - Make sure the `no-unused-vars` rule is enabled in
   [ESLint](https://eslint.org/docs/rules/no-unused-vars) and any imported plugin
   are actually being loaded.
+- In case you've the `@typescript-eslint/no-floating-promises` enabled,
+please double-check that your `.eslintrc` includes `allowForKnownSafePromises`
```
Comment (JoshuaKGoldberg) [inline docs/Reference/TypeScript.md:665]: Now that [ESLint's new flat config](https://eslint.org/blog/2022/08/new-config-system-part-2) is the default, `.eslintrc` is no longer what many repos will use instead of `eslint.config.*`:

```suggestion
please double-check that your ESLint configuration includes a `allowForKnownSafePromises`
```

**Review by JoshuaKGoldberg — APPROVED** [review]
Looks great to me, thanks for including me in the discussion! ❤️ 

Per the long thread, I'm inferring `FastifyInstance` might also want to be in the suggested list? Otherwise this all looks great to me

**Review by mcollina — APPROVED** [review]
lgtm
